### PR TITLE
Chore: Clean up things when parents are deleted

### LINKED ISSUE #40: Handle delete related objects
Requires thread delete: https://github.com/textileio/go-threads/issues/176

### LINKED ISSUE #175: Destroy entire account API/cmd

### LINKED ISSUE #176: Destroy user and data API/cmd

### PR DESCRIPTION
**Pull request by sanderpick**
Closes #40 
Closes #175 
Closes #176 

This mostly revolves around all the things that have to be done when deleting an account.

### CODE REVIEW

File: api/buckets/service.go [hunk 1]
```diff
@@ -47,9 +42,10 @@ type Service struct {
 	IPFSClient     iface.CoreAPI
 	FilecoinClient *fc.Client
 
-	GatewayUrl  string
+	GatewayUrl string
+
 	DNSManager  *dns.Manager
-	IPNSManager *IPNSManager
+	IPNSManager *ipns.Manager
```
Comment (sanderpick) [inline api/buckets/service.go:48]: Manager got pulled into an `ipns` package.

File: api/hub/service.go [hunk 2]
```diff
@@ -436,3 +461,74 @@ func ownerFromContext(ctx context.Context) crypto.PubKey {
 	}
 	return org.Key
 }
+
+func (s *Service) destroyAccount(ctx context.Context, a *c.Account) error {
```
Comment (sanderpick) [inline api/hub/service.go:465]: Here's the meat of destroying a dev / org account.

File: api/users/pb/users.proto [hunk 3]
```diff
@@ -19,6 +19,7 @@ message GetThreadRequest {
 message GetThreadReply {
     bytes ID = 1;
     string name = 2;
+    bool isDB = 3;
```
Comment (sanderpick) [inline api/users/pb/users.proto:22]: This is needed to differentiate between thread IDs for DBs, which can be used for buckets, or raw / log type thread IDs.

File: cmd/tt/bucket.go [hunk 4]
```diff
@@ -30,7 +30,7 @@ func init() {
 	initBucketPathCmd.PersistentFlags().String("org", "", "Org username")
 	initBucketPathCmd.PersistentFlags().Bool("public", false, "Allow public access")
 	initBucketPathCmd.PersistentFlags().String("thread", "", "Thread ID")
-	initBucketPathCmd.Flags().Bool("existing", false, "If set, initalizes from an existing remote Bucket")
+	initBucketPathCmd.Flags().Bool("existing", false, "If set, initalizes from an existing remote bucket")
```
Comment (sanderpick) [inline cmd/tt/bucket.go:33]: Just lowercasing bucket to be consistent

File: cmd/tt/destroy.go [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+	"github.com/manifoldco/promptui"
+	"github.com/spf13/cobra"
+	"github.com/textileio/textile/cmd"
+)
+
+func init() {
+	rootCmd.AddCommand(destroyCmd)
+}
+
+var destroyCmd = &cobra.Command{
```
Comment (sanderpick) [inline cmd/tt/destroy.go:16]: New command `destroy` requires confirmation.

File: core/core.go [hunk 6]
```diff
@@ -510,7 +526,29 @@ func (t *Textile) threadInterceptor() grpc.UnaryServerInterceptor {
...
+			deleteID, err = thread.Cast(req.(*dbpb.DeleteDBRequest).DbID)
+			if err != nil {
+				return nil, err
+			}
+			keys, err := t.co.IPNSKeys.ListByThreadID(ctx, deleteID)
+			if err != nil {
+				return nil, err
+			}
+			if len(keys) != 0 {
+				return nil, status.Error(codes.FailedPrecondition, "DB not empty. Delete buckets first.")
```
Comment (sanderpick) [inline core/core.go:547]: A condition I came up with that feels reasonable: You cannot delete a database without first deleting all buckets, _unless_ you're destroying your entire account.

File: ipns/ipns.go [hunk 7]
```diff
@@ -0,0 +1,170 @@
+package ipns
```
Comment (sanderpick) [inline ipns/ipns.go:1]: The new `ipns` package handles all keys and publishing.

File: api/buckets/buckets.go [hunk 8]
```diff
@@ -12,7 +12,7 @@ import (
 	db "github.com/textileio/go-threads/db"
 )
 
-const cname = "buckets"
+const CollectionName = "buckets"
```
Comment (jsign) [inline api/buckets/buckets.go:15]: I did this same change on my opened PR :P So agree!

File: collections/ipnskeys.go [hunk 9]
```diff
@@ -65,6 +68,29 @@ func (k *IPNSKeys) GetByCid(ctx context.Context, cid string) (*IPNSKey, error) {
 	return decodeIPNSKey(raw)
 }
 
+func (k *IPNSKeys) ListByThreadID(ctx context.Context, threadID thread.ID) ([]IPNSKey, error) {
+	cursor, err := k.col.Find(ctx, bson.M{"thread_id": threadID.Bytes()})
```
Comment (jsign) [inline collections/ipnskeys.go:72]: It seems that `cursor` has a Close function.
Should this be deferred?
Might apply to more places if that's the case.

Comment (sanderpick) [reply collections/ipnskeys.go:72]: Ah yeah, good eye. Fixing.

**Pull request merged**